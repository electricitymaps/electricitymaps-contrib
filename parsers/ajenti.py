"""
Initial PR https://github.com/electricitymaps/electricitymaps-contrib/pull/2456
Discussion thread https://github.com/electricitymaps/electricitymaps-contrib/issues/636
A promotion webpage for King's Island energy production is here: https://www.hydro.com.au/clean-energy/hybrid-energy-solutions/success-stories/king-island
As of 09/2020, it embeds with <iframe> the URI https://data.ajenti.com.au/KIREIP/index.html

About the data, the feed we get seems to be counters with a 2 seconds interval.
That means that if we fetch these counters every 15 minutes, we only are reading "instantaneous" metters that could differ from the total quantity of energies at play.
To get the very exact data, we would need to have a parser running constanty to collect those 2-sec interval counters.
"""

import json
from datetime import datetime
from logging import Logger, getLogger
from zoneinfo import ZoneInfo

from requests import Session
from signalr import Connection

ZONE_PARAMS = {
    # King Island
    # https://www.hydro.com.au/clean-energy/hybrid-energy-solutions/success-stories/king-island
    "AU-TAS-KI": {
        "hub": "TagHub",
        "method": "Dashboard",
        "tz": "Australia/Currie",
        "source": "hydro.com.au",  # Iframe: https://data.ajenti.com.au/KIREIP/index.html
    },
    # Flinders Island
    # https://github.com/electricitymaps/electricitymaps-contrib/issues/2533
    # https://en.wikipedia.org/wiki/Flinders_Island
    # https://www.hydro.com.au/clean-energy/hybrid-energy-solutions/success-stories/flinders-island
    "AU-TAS-FI": {
        "hub": "flindershub",
        "method": "SendDashboard",
        "tz": "Australia/Hobart",
        "source": "hydro.com.au",
    },
    # Rottnest Island
    # https://github.com/electricitymaps/electricitymaps-contrib/issues/2534
    # https://en.wikipedia.org/wiki/Rottnest_Island
    # https://www.hydro.com.au/clean-energy/hybrid-energy-solutions/success-stories/rottnest-island
    "AU-WA-RI": {
        "hub": "HogsHub",
        "method": "SendDashboard",
        "tz": "Australia/Perth",
        "source": "hydro.com.au",
    },
}


class SignalR:
    def __init__(self, url):
        self.url = url

    def update_res(self, msg):
        if msg != {}:
            self.res = msg

    def get_value(self, hub, method):
        self.res = {}
        with Session() as session:
            # create a connection
            connection = Connection(self.url, session)
            chat = connection.register_hub(hub)
            chat.client.on(method, self.update_res)
            connection.start()
            connection.wait(3)
            connection.close()
            return self.res


def parse_payload(logger: Logger, payload) -> dict:
    technologies_parsed = {
        "biomass": 0,
        "battery": 0,
        "coal": 0,
        "flywheel": 0,
        "gas": 0,
        "hydro": 0,
        "nuclear": 0,
        "oil": 0,
        "solar": 0,
        "wind": 0,
        "geothermal": 0,
        "unknown": 0,
    }
    if "technologies" not in payload:
        raise KeyError(
            f"No 'technologies' in payload\n" f"serie : {json.dumps(payload)}"
        )
    else:
        logger.debug(f"serie : {json.dumps(payload)}")
    for technology in payload["technologies"]:
        # rename upstream key to match our payload needs
        if technology["id"] == "diesel":
            technology["id"] = "oil"

        assert technology["unit"] == "kW"
        # The upstream API gives us kW, we need MW
        technologies_parsed[technology["id"]] = int(technology["value"]) / 1000

    # King's Island uses some percentage of biodiesel; since we account differently traditionnal oil & biomass, here we do math to separate both
    if "biodiesel" in payload:
        # sanity check to be sure that we are not remaking data that the data source could otherwise give us
        assert "biomass" and "oil" != 0
        technologies_parsed["biomass"] = (
            technologies_parsed["oil"] * payload["biodiesel"]["percent"] / 100
        )
        technologies_parsed["oil"] = (
            technologies_parsed["oil"] * (100 - payload["biodiesel"]["percent"]) / 100
        )

    logger.debug(f"production : {json.dumps(technologies_parsed)}")

    return technologies_parsed


# Both keys battery and flywheel are negative when storing energy, and positive when feeding energy to the grid
def format_storage_techs(technologies_parsed):
    # sometimes we don't get those keys. For instance Rottnest island don't have any storage.
    if "battery" and "flywheel" in technologies_parsed:
        storage_techs = technologies_parsed["battery"] + technologies_parsed["flywheel"]
    else:
        storage_techs = 0
    _battery_production = (
        storage_techs if storage_techs > 0 else 0
    )  # TODO: Should this just be removed?
    _battery_storage = storage_techs if storage_techs < 0 else 0


def sum_storage_techs(technologies_parsed):
    storage_techs = technologies_parsed["battery"] + technologies_parsed["flywheel"]

    return storage_techs


def fetch_production(
    zone_key: str = "AU-TAS-KI",
    session: Session | None = None,
    target_datetime=None,
    logger: Logger = getLogger(__name__),
) -> dict:
    if target_datetime is not None:
        raise NotImplementedError(
            "The datasource currently implemented is only real time"
        )

    # get the specific zone parameters
    try:
        hub, dashboard, tz, source = (
            ZONE_PARAMS[zone_key]["hub"],
            ZONE_PARAMS[zone_key]["method"],
            ZONE_PARAMS[zone_key]["tz"],
            ZONE_PARAMS[zone_key]["source"],
        )
    except KeyError as e:
        raise KeyError("The zone " + zone_key + " isn't implemented") from e

    payload = SignalR("https://data.ajenti.com.au/live/signalr").get_value(
        hub, dashboard
    )
    technologies_parsed = parse_payload(logger, payload)
    storage_techs = sum_storage_techs(technologies_parsed)

    return {
        "zoneKey": zone_key,
        "datetime": datetime.now(tz=ZoneInfo(tz)),
        "production": {
            "biomass": technologies_parsed["biomass"],
            "coal": technologies_parsed["coal"],
            "gas": technologies_parsed["gas"],
            "hydro": technologies_parsed["hydro"],
            "nuclear": technologies_parsed["nuclear"],
            "oil": technologies_parsed["oil"],
            "solar": technologies_parsed["solar"],
            "wind": 0
            if technologies_parsed["wind"] < 0 and technologies_parsed["wind"] > -0.1
            else technologies_parsed[
                "wind"
            ],  # If wind between 0 and -0.1 set to 0 to ignore self-consumption
            "geothermal": technologies_parsed["geothermal"],
            "unknown": technologies_parsed["unknown"],
        },
        "storage": {
            "battery": storage_techs
            * -1  # Somewhat counterintuitively,to ElectricityMap positive means charging and negative means discharging
        },
        "source": source,
    }


if __name__ == "__main__":
    print(fetch_production())
