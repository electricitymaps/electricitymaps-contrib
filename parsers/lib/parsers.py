import importlib

from electricitymap.contrib.config import EXCHANGES_CONFIG, ZONES_CONFIG

# Prepare all parsers
CONSUMPTION_PARSERS = {}
PRODUCTION_PARSERS = {}
PRODUCTION_PER_MODE_FORECAST_PARSERS = {}
PRODUCTION_PER_UNIT_PARSERS = {}
EXCHANGE_PARSERS = {}
PRICE_PARSERS = {}
CONSUMPTION_FORECAST_PARSERS = {}
GENERATION_FORECAST_PARSERS = {}
EXCHANGE_FORECAST_PARSERS = {}
PRODUCTION_CAPACITY_PARSERS = {}
DAYAHEAD_LOCATIONAL_MARGINAL_PRICE_PARSERS = {}
REALTIME_LOCATIONAL_MARGINAL_PRICE_PARSERS = {}

PARSER_KEY_TO_DICT = {
    "consumption": CONSUMPTION_PARSERS,
    "production": PRODUCTION_PARSERS,
    "productionPerUnit": PRODUCTION_PER_UNIT_PARSERS,
    "productionPerModeForecast": PRODUCTION_PER_MODE_FORECAST_PARSERS,
    "exchange": EXCHANGE_PARSERS,
    "price": PRICE_PARSERS,
    "dayaheadLocationalMarginalPrice": DAYAHEAD_LOCATIONAL_MARGINAL_PRICE_PARSERS,
    "realtimeLocationalMarginalPrice": REALTIME_LOCATIONAL_MARGINAL_PRICE_PARSERS,
    "consumptionForecast": CONSUMPTION_FORECAST_PARSERS,
    "generationForecast": GENERATION_FORECAST_PARSERS,
    "exchangeForecast": EXCHANGE_FORECAST_PARSERS,
    "productionCapacity": PRODUCTION_CAPACITY_PARSERS,
}


def _parser_key_to_parser_folder(parser_key: str):
    return (
        "electricitymap.contrib.capacity_parsers"
        if parser_key == "productionCapacity"
        else "parsers"
    )


# Read all zones
for zone_id, zone_config in ZONES_CONFIG.items():
    for parser_key, v in zone_config.get("parsers", {}).items():
        mod_name, fun_name = v.split(".")
        mod = importlib.import_module(
            f"{_parser_key_to_parser_folder(parser_key)}.{mod_name}"
        )
        PARSER_KEY_TO_DICT[parser_key][zone_id] = getattr(mod, fun_name)


# Read all exchanges
for exchange_id, exchange_config in EXCHANGES_CONFIG.items():
    for parser_key, v in exchange_config.get("parsers", {}).items():
        mod_name, fun_name = v.split(".")
        mod = importlib.import_module(f"parsers.{mod_name}")
        PARSER_KEY_TO_DICT[parser_key][exchange_id] = getattr(mod, fun_name)
