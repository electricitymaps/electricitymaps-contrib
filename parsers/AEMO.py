import io
import re
import zipfile
from datetime import datetime
from logging import Logger, getLogger
from typing import Any
from urllib.parse import urljoin
from zoneinfo import ZoneInfo

import pandas as pd
from bs4 import BeautifulSoup
from requests import Session

from electricitymap.contrib.lib.models.event_lists import TotalConsumptionList
from electricitymap.contrib.lib.models.events import EventSourceType
from electricitymap.contrib.lib.types import ZoneKey

SOURCE = "aemo.com.au"

ZONE_KEY_TO_REGION = {
    "AU-NSW": "NSW1",
    "AU-QLD": "QLD1",
    "AU-SA": "SA1",
    "AU-TAS": "TAS1",
    "AU-VIC": "VIC1",
    "AU-WA": "WEM",  # This zone is not implemented yet
}

ZONE_KEY_TO_TIMEZONE = {
    "AU-NSW": ZoneInfo("Australia/Sydney"),
    "AU-QLD": ZoneInfo("Australia/Brisbane"),
    "AU-SA": ZoneInfo("Australia/Adelaide"),
    "AU-TAS": ZoneInfo("Australia/Hobart"),
    "AU-VIC": ZoneInfo("Australia/Melbourne"),
    "AU-WA": ZoneInfo("Australia/Perth"),  # This zone is not implemented yet
}

# TODO, what about the other zone in Australia (AU-WA)? Check remaining zone


def find_document(session, target_datetime):
    # Fetch the directory listing
    base_url = "http://nemweb.com.au/Reports/CURRENT/Operational_Demand/FORECAST_HH/"
    response = session.get(base_url)

    # Parse with BeautifulSoup
    soup = BeautifulSoup(response.text, "html.parser")

    # String datetime
    target_date_str = target_datetime.strftime("%Y%m%d")

    # Find matching links
    matching_links = soup.find_all(
        "a",
        href=re.compile(
            rf"PUBLIC_FORECAST_OPERATIONAL_DEMAND_HH_{target_date_str}\d+_\d+\.zip"
        ),
    )

    if matching_links:
        target_link = matching_links[-1]  # Get the last (most recent) link
        full_url = urljoin(base_url, target_link["href"])  # Construct full URL

        file_response = session.get(full_url)

        with zipfile.ZipFile(io.BytesIO(file_response.content)) as z:
            csv_filename = z.namelist()[0]
            with z.open(csv_filename) as f:
                df = pd.read_csv(f)
                return df
    else:
        print("No matching files found")


def fetch_consumption_forecast(
    zone_key: ZoneKey,  # "AU-NSW", "AU-QLD", "AU-SA", "AU-TAS", "AU-VIC", "AU-WA"
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict[str, Any]]:
    """Consumption forecast in MW every half an hour for 10 days ahead.
    Only for NSW1, QND1, SA1, TAS1, VIC1 zones."""
    session = session or Session()

    if target_datetime is None:
        target_datetime = datetime.now(tz=ZONE_KEY_TO_TIMEZONE[zone_key])

    df = find_document(session, target_datetime)

    # Transform dataframe
    df.columns = df.iloc[0]
    df = df.iloc[1:-1].reset_index(drop=True)

    #
    region = ZONE_KEY_TO_REGION.get(zone_key)
    all_consumption_events = df[
        df["REGIONID"] == region
    ]  # all events with a datetime and a consumption value
    consumption_list = TotalConsumptionList(logger)
    for _, event in all_consumption_events.iterrows():
        datetime_object = datetime.strptime(
            event["INTERVAL_DATETIME"], "%Y/%m/%d %H:%M:%S"
        ).replace(tzinfo=ZONE_KEY_TO_TIMEZONE[zone_key])

        consumption_list.append(
            zoneKey=zone_key,
            datetime=datetime_object,
            consumption=float(
                event["OPERATIONAL_DEMAND_POE50"]
            ),  # 50% probability of exceedance operational demand forecast value
            source=SOURCE,
            sourceType=EventSourceType.forecasted,
        )
    return consumption_list.to_list()


if __name__ == "__main__":
    """Main method, never used by the electricityMap backend, but handy for testing."""

    print(fetch_consumption_forecast("AU-NSW"))
    print(fetch_consumption_forecast("AU-QLD"))
    print(fetch_consumption_forecast("AU-SA"))
    print(fetch_consumption_forecast("AU-TAS"))
    print(fetch_consumption_forecast("AU-VIC"))
    print(
        fetch_consumption_forecast("AU-WA")
    )  # Not implemented yet. It returns an empty list
