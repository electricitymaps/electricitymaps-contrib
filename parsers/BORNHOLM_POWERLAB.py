#!/usr/bin/env python3
from datetime import datetime
from logging import Logger, getLogger
from zoneinfo import ZoneInfo

from requests import Session

from electricitymap.contrib.lib.models.event_lists import (
    ExchangeList,
    ProductionBreakdownList,
)
from electricitymap.contrib.lib.models.events import ProductionMix
from electricitymap.contrib.lib.types import ZoneKey

PRODUCTION_MAPPING = {
    "wind": "wind_turbines",
    "biomass": "factory",
    "solar": "solar_cells",
}
LATEST_DATA_URL = "https://bornholm.powerlab.dk/visualizer/latestdata"

SOURCE = "bornholm.powerlab.dk"
TIMEZONE = ZoneInfo("UTC")


def fetch_production(
    zone_key: ZoneKey = ZoneKey("DK-BHM"),
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    """Requests the last known production mix (in MW) of a given country."""
    if target_datetime:
        raise NotImplementedError("This parser is not yet able to parse past dates")
    if session is None:
        session = Session()
    response = session.get(LATEST_DATA_URL).json()
    production = ProductionBreakdownList(logger)
    production.append(
        zoneKey=zone_key,
        datetime=datetime.fromtimestamp(response["latest"], tz=TIMEZONE),
        source=SOURCE,
        production=ProductionMix(
            biomass=response["sub"]["factory"],
            solar=response["sub"]["solar_cells"],
            wind=response["sub"]["wind_turbines"],
        ),
    )
    return production.to_list()


def fetch_exchange(
    zone_key1: ZoneKey = ZoneKey("DK-BHM"),
    zone_key2: ZoneKey = ZoneKey("SE-SE4"),
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    """Requests the last known power exchange (in MW) between two countries."""
    if target_datetime:
        raise NotImplementedError("This parser is not yet able to parse past dates")
    if session is None:
        session = Session()
    response = session.get(LATEST_DATA_URL).json()
    exchange = ExchangeList(logger)
    sorted_zone_keys = sorted([zone_key1, zone_key2])
    direction = 1 if zone_key1 == sorted_zone_keys[0] else -1
    exchange.append(
        zoneKey=ZoneKey("->".join(sorted_zone_keys)),
        datetime=datetime.fromtimestamp(response["latest"], tz=TIMEZONE),
        source=SOURCE,
        netFlow=direction * response["sub"]["seacable"],
    )

    return exchange.to_list()


if __name__ == "__main__":
    """Main method, never used by the Electricity Map backend, but handy for testing."""

    print("fetch_production() ->")
    print(fetch_production())
    print("fetch_exchange(DK-BHM, SE-SE4) ->")
    print(fetch_exchange())
