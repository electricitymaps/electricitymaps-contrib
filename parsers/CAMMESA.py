#!/usr/bin/env python3


from datetime import datetime
from logging import Logger, getLogger
from zoneinfo import ZoneInfo

import pandas as pd
from requests import Session

from electricitymap.contrib.lib.models.event_lists import (
    ExchangeList,
    ProductionBreakdownList,
)
from electricitymap.contrib.lib.models.events import ProductionMix
from electricitymap.contrib.lib.types import ZoneKey
from parsers.lib.exceptions import ParserException

# Useful links.
# https://en.wikipedia.org/wiki/Electricity_sector_in_Argentina
# https://en.wikipedia.org/wiki/List_of_power_stations_in_Argentina
# http://globalenergyobservatory.org/countryid/10#
# http://www.industcards.com/st-other-argentina.htm

# API Documentation: https://api.cammesa.com/demanda-svc/swagger-ui.html
CAMMESA_DEMANDA_ENDPOINT = (
    "https://api.cammesa.com/demanda-svc/generacion/ObtieneGeneracioEnergiaPorRegion/"
)
CAMMESA_EXCHANGE_ENDPOINT = (
    "https://api.cammesa.com/demanda-svc/demanda/IntercambioCorredoresGeo/"
)
CAMMESA_RENEWABLES_ENDPOINT = "https://cdsrenovables.cammesa.com/exhisto/RenovablesService/GetChartTotalTRDataSource/"

# Expected arrow direction if flow is first -> second
EXCHANGE_NAME_DIRECTION_MAPPING = {
    "AR->CL-SEN": ("ARG-CHI", 180),
    "AR->PY": ("ARG-PAR", 45),
    "AR->UY": ("ARG-URU", 0),
    "AR-NEA->BR-S": ("ARG-BRA", 45),
    "AR-BAS->AR-COM": ("PBA-COM", 225),
    "AR-CEN->AR-COM": ("CEN-COM", 270),
    "AR-CEN->AR-NOA": ("CEN-NOA", 135),
    "AR-CUY->AR-COM": ("CUY-COM", 315),
    "AR-CEN->AR-CUY": ("CEN-CUY", 225),
    "AR-LIT->AR-NEA": ("LIT-NEA", 45),
    "AR-BAS->AR-LIT": ("PBA-LIT", 90),
    "AR-CUY->AR-NOA": ("CUY-NOA", 45),
    "AR-LIT->AR-CEN": ("LIT-CEN", 225),
    "AR-LIT->AR-NOA": ("LIT-NOA", 135),
    "AR-BAS->AR-CEN": ("PBA-CEN", 135),
    "AR-COM->AR-PAT": ("COM-PAT", 270),
    "AR-NEA->AR-NOA": ("NEA-NOA", 180),
}

SOURCE = "cammesaweb.cammesa.com"

SPLIT_UNKNOWN_PRODUCTION = {  # To be updated with the latest data when available
    2017: {"oil": 0.12, "gas": 0.85, "coal": 0.03},
    2018: {"oil": 0.07, "gas": 0.90, "coal": 0.03},
    2019: {"oil": 0.04, "gas": 0.95, "coal": 0.01},
    2020: {"oil": 0.075, "gas": 0.90, "coal": 0.025},
    2021: {"oil": 0.12, "gas": 0.85, "coal": 0.03},
    2022: {"oil": 0.17, "gas": 0.80, "coal": 0.03},
}  # source: EIA 2022 (https://www.iea.org/data-and-statistics/data-tools/energy-statistics-data-browser?country=ARGENTINA&fuel=Energy%20supply&indicator=ElecGenByFuel)
df_split_unknown_production = pd.DataFrame(SPLIT_UNKNOWN_PRODUCTION).T
df_split_unknown_production = df_split_unknown_production.reindex(
    range(2017, datetime.now().year + 1), method="ffill"
)


def fetch_production(
    zone_key: ZoneKey = ZoneKey("AR"),
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    """Requests up to date list of production mixes (in MW) of a given country."""

    if target_datetime:
        raise NotImplementedError("This parser is not yet able to parse past dates")

    current_session = session or Session()

    conventional_production = non_renewables_production_mix(
        zone_key, current_session, logger
    )
    renewables_production = renewables_production_mix(zone_key, current_session, logger)
    # Hydro comes from both conventional and renewables production which are merged together
    return ProductionBreakdownList.merge_production_breakdowns(
        [conventional_production, renewables_production],
        logger,
        matching_timestamps_only=True,
    ).to_list()


def renewables_production_mix(
    zone_key: ZoneKey, session: Session, logger: Logger
) -> ProductionBreakdownList:
    """Retrieves production mix for renewables using CAMMESA's API"""

    today = (
        datetime.now()
        .astimezone(tz=ZoneInfo("America/Argentina/Buenos_Aires"))
        .strftime("%d-%m-%Y")
    )
    params = {"desde": today, "hasta": today}
    renewables_response = session.get(CAMMESA_RENEWABLES_ENDPOINT, params=params)
    assert renewables_response.status_code == 200, (
        "Exception when fetching production for "
        f"{zone_key}: error when calling url={CAMMESA_RENEWABLES_ENDPOINT} with payload={params}"
    )

    production_list = renewables_response.json()
    renewables_production = ProductionBreakdownList(logger)
    for production_info in production_list:
        renewables_production.append(
            zoneKey=zone_key,
            datetime=datetime.strptime(
                production_info["momento"], "%Y-%m-%dT%H:%M:%S.%f%z"
            ),
            production=ProductionMix(
                biomass=production_info["biocombustible"],
                hydro=production_info["hidraulica"],
                solar=production_info["fotovoltaica"],
                wind=production_info["eolica"],
            ),
            source=SOURCE,
        )

    return renewables_production


def non_renewables_production_mix(
    zone_key: ZoneKey, session: Session, logger: Logger
) -> ProductionBreakdownList:
    """Retrieves production mix for non renewables using CAMMESA's API"""

    params = {"id_region": 1002}
    api_cammesa_response = session.get(CAMMESA_DEMANDA_ENDPOINT, params=params)
    assert api_cammesa_response.status_code == 200, (
        "Exception when fetching production for "
        f"{zone_key}: error when calling url={CAMMESA_DEMANDA_ENDPOINT} with payload={params}"
    )
    production_list = api_cammesa_response.json()
    conventional_production = ProductionBreakdownList(logger)
    for production_info in production_list:
        # Convert fecha string to datetime and extract the year
        production_datetime = datetime.strptime(
            production_info["fecha"], "%Y-%m-%dT%H:%M:%S.%f%z"
        )
        production_year = production_datetime.year

        # use the split from EIA 2022 to split the termico production into gas and oil
        for mode in ["gas", "oil", "coal"]:
            production_info[mode] = (
                production_info["termico"]
                * df_split_unknown_production.loc[production_year][mode]
            )  # use the split from EIA 2022 to split the termico production into gas and oil
        conventional_production.append(
            zoneKey=zone_key,
            datetime=production_datetime,
            production=ProductionMix(
                hydro=production_info["hidraulico"],
                nuclear=production_info["nuclear"],
                # As of 2022 thermal energy is mostly natural gas but
                # the data is not split. We put it into unknown for now.
                # More info: see page 21 in https://microfe.cammesa.com/static-content/CammesaWeb/download-manager-files/Sintesis%20Mensual/Informe%20Mensual_2021-12.pdf
                # unknown=production_info["termico"],
                gas=production_info["gas"],
                oil=production_info["oil"],
                coal=production_info["coal"],
            ),
            source=SOURCE,
        )

    return conventional_production


def fetch_exchange(
    zone_key1: ZoneKey,
    zone_key2: ZoneKey,
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    """Requests the last known power exchange (in MW) between two zones."""
    if target_datetime:
        raise NotImplementedError("This parser is not yet able to parse past dates")

    sorted_zone_keys = ZoneKey("->".join(sorted([zone_key1, zone_key2])))
    if sorted_zone_keys not in EXCHANGE_NAME_DIRECTION_MAPPING:
        raise ParserException(
            parser="CAMMESA.py",
            message="This exchange is not currently implemented",
            zone_key=sorted_zone_keys,
        )

    current_session = session or Session()

    api_cammesa_response = current_session.get(CAMMESA_EXCHANGE_ENDPOINT)
    if not api_cammesa_response.ok:
        raise ParserException(
            parser="CAMMESA.py",
            message=f"Exception when fetching exchange for {sorted_zone_keys}: error when calling url={CAMMESA_EXCHANGE_ENDPOINT}",
            zone_key=sorted_zone_keys,
        )
    exchange_name, expected_angle = EXCHANGE_NAME_DIRECTION_MAPPING[sorted_zone_keys]
    exchange_list = api_cammesa_response.json()["features"]
    exchange_data = next(
        (
            exchange["properties"]
            for exchange in exchange_list
            if exchange["properties"]["nombre"] == exchange_name
        ),
        None,
    )
    if exchange_data is None:
        raise ParserException(
            parser="CAMMESA.py",
            message=f"Exception when fetching exchange for {sorted_zone_keys}: exchange not found",
            zone_key=sorted_zone_keys,
        )
    given_angle = int(exchange_data["url"][6:])
    flow = int(exchange_data["text"])

    # inverse flow if arrow doesn't match expected direction
    if expected_angle != given_angle:
        flow = -flow
    exchange_datetime = datetime.fromisoformat(
        exchange_data["fecha"][:-2] + ":" + exchange_data["fecha"][-2:]
    )
    exchanges = ExchangeList(logger)
    exchanges.append(
        zoneKey=sorted_zone_keys,
        datetime=exchange_datetime,
        netFlow=flow,
        source=SOURCE,
    )
    return exchanges.to_list()


def fetch_price(
    zone_key: str = "AR",
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> dict:
    """Requests the last known power price of a given country."""

    raise NotImplementedError("Fetching the price is not currently implemented")


if __name__ == "__main__":
    """Main method, never used by the Electricity Map backend, but handy for testing."""

    print("fetch_production() ->")
    print(fetch_production())
    print("fetch_exchange(AR, PY) ->")
    print(fetch_exchange(ZoneKey("AR"), ZoneKey("PY")))
    print("fetch_exchange(AR, UY) ->")
    print(fetch_exchange(ZoneKey("AR"), ZoneKey("UY")))
    print("fetch_exchange(AR, CL-SEN) ->")
    print(fetch_exchange(ZoneKey("AR"), ZoneKey("CL-SEN")))
    print("fetch_exchange(AR-CEN, AR-COM) ->")
    print(fetch_exchange(ZoneKey("AR-CEN"), ZoneKey("AR-COM")))
    print("fetch_exchange(AR-BAS, AR-LIT) ->")
    print(fetch_exchange(ZoneKey("AR-BAS"), ZoneKey("AR-LIT")))
    print("fetch_exchange(AR-COM, AR-PAT) ->")
    print(fetch_exchange(ZoneKey("AR-COM"), ZoneKey("AR-PAT")))
