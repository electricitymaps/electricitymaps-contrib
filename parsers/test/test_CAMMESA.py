import json
from importlib import resources

import pytest
from requests_mock import ANY, GET

from electricitymap.contrib.lib.types import ZoneKey
from parsers.CAMMESA import (
    CAMMESA_DEMANDA_ENDPOINT,
    CAMMESA_RENEWABLES_ENDPOINT,
    fetch_exchange,
    fetch_production,
)


@pytest.fixture(autouse=True)
def mock_response(adapter):
    adapter.register_uri(
        GET,
        ANY,
        json=json.loads(
            resources.files("parsers.test.mocks.Cammesa")
            .joinpath("exchanges.json")
            .read_text()
        ),
    )
    adapter.register_uri(
        GET,
        CAMMESA_DEMANDA_ENDPOINT,
        json=json.loads(
            resources.files("parsers.test.mocks.Cammesa")
            .joinpath("conventional_production.json")
            .read_text()
        ),
    )
    adapter.register_uri(
        GET,
        CAMMESA_RENEWABLES_ENDPOINT,
        json=json.loads(
            resources.files("parsers.test.mocks.Cammesa")
            .joinpath("renewable_production.json")
            .read_text()
        ),
    )


def test_exchanges_AR_CL_SEN(session, snapshot):
    assert snapshot == fetch_exchange(
        ZoneKey("AR"),
        ZoneKey("CL-SEN"),
        session=session,
    )


def test_exchanges_AR_BAS_AR_COM(session, snapshot):
    assert snapshot == fetch_exchange(
        ZoneKey("AR-BAS"),
        ZoneKey("AR-COM"),
        session=session,
    )


def test_fetch_production(session, snapshot):
    assert snapshot == fetch_production(ZoneKey("AR"), session=session)
