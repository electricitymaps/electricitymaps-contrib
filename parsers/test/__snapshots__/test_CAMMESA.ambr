# serializer version: 1
# name: test_exchanges_AR_BAS_AR_COM
  list([
  ])
# ---
# name: test_exchanges_AR_CL_SEN
  list([
    dict({
      'datetime': datetime.datetime(2023, 8, 30, 2, 45, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=75600))),
      'netFlow': 200.0,
      'sortedZoneKeys': 'AR->CL-SEN',
      'source': 'cammesaweb.cammesa.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
    }),
  ])
# ---
# name: test_fetch_production
  list([
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2023, 9, 7, 0, 20, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=75600))),
      'production': dict({
        'biomass': 161.85,
        'hydro': 6086.5,
        'nuclear': 1388.4,
        'solar': 0.0,
        'unknown': 7487.1,
        'wind': 2009.54,
      }),
      'source': 'cammesaweb.cammesa.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'AR',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2023, 9, 7, 0, 45, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=75600))),
      'production': dict({
        'biomass': 157.2,
        'hydro': 5893.23,
        'nuclear': 1390.4,
        'solar': 0.0,
        'unknown': 7223.0,
        'wind': 1944.1,
      }),
      'source': 'cammesaweb.cammesa.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'AR',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2023, 9, 7, 0, 50, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=75600))),
      'production': dict({
        'biomass': 158.55,
        'hydro': 6028.12,
        'nuclear': 1394.0,
        'solar': 0.0,
        'unknown': 7206.0,
        'wind': 1922.51,
      }),
      'source': 'cammesaweb.cammesa.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'AR',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2023, 9, 7, 1, 45, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=75600))),
      'production': dict({
        'biomass': 148.3,
        'hydro': 6592.0,
        'nuclear': 1394.6,
        'solar': 0.0,
        'unknown': 6802.2,
        'wind': 1900.64,
      }),
      'source': 'cammesaweb.cammesa.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'AR',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2023, 9, 7, 1, 50, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=75600))),
      'production': dict({
        'biomass': 149.74,
        'hydro': 6678.26,
        'nuclear': 1387.0,
        'solar': 0.0,
        'unknown': 6758.0,
        'wind': 1871.5,
      }),
      'source': 'cammesaweb.cammesa.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'AR',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2023, 9, 7, 1, 55, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=75600))),
      'production': dict({
        'biomass': 155.96,
        'hydro': 6525.16,
        'nuclear': 1388.1,
        'solar': 0.0,
        'unknown': 6791.2,
        'wind': 1835.68,
      }),
      'source': 'cammesaweb.cammesa.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'AR',
    }),
  ])
# ---
